src/org/exist/storage/XQueryPool.java

```diff
@@ -146,23 +147,19 @@ private void returnObject(final Source source, final CompiledXQuery xquery) {
             if (pool.size() < maxPoolSize) {
                 source.setCacheTimestamp(System.currentTimeMillis());
 
-                final Deque<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ArrayDeque<>());
+                final Set<CompiledXQuery> stack = pool.computeIfAbsent(source, (k) -> new ConcurrentSkipListSet<>());
 
                 if (stack.size() < maxStackSize) {
-                    // check if the query is already in pool before adding,
-                    // may happen for modules, don't add it a second time!
-                    if(!stack.contains(xquery)) {
-                        stack.push(xquery);
-                    }
+                    stack.add(xquery);
```

Comment:
I do not see this as problem. `timeoutCheck` removing `valid compiled queries` too.

The worse case for the problem you describe that some valid compiled queries will be GCed and it will not be reused. I do not see it as critical issue, more to that I do not see this as issue at all. That is some trade for better concurrency.

There are always chance that compiled query during evaluation become invalid, or modules changed during compilation.

Anyway, all that cases more develop sites issue, then production one, because on production side scripts do not change too often.